### ISSUE TITLE
Primitive type non-nullable as required filter?

### ISSUE THREAD

**Issue by ThisNoName** [domaindrivendev/Swashbuckle.AspNetCore#498]
I know swagger specification requires all required fields be marked required.  But really wish there's an option to avoid some serious tendonitis here ^_^

In the pre-core Swashbuckle (5.5.3?), there was an [AutoRestExtension ](https://github.com/Alegrowin/Swashbuckle.AutoRestExtensions) that generate swagger schema based on primitive type, like

    int IntRequired {get;set;}
    int? IntOptional {get;set;}

When I try to port it to core, I can't seem to get the underlying nullable types. The nonNullableProperties always end up empty. Any thoughts? Or better, someone did it already ... Thanks

    public class NonNullableAsRequiredSchemaFilter : ISchemaFilter
    {
    public void Apply(Schema schema, SchemaFilterContext context)
    {
        if (schema.Properties != null)
        {
            // Find properties with non-nullable types
            var nonNullableProperties = schema.Properties
                .Where(x => x.Value.HasNullable() && !x.Value.IsNullable())
                .Select(x => x.Key)
                .ToArray();

            if (nonNullableProperties.Length > 0)
            {
                // Mark them as required
                if (schema.Required == null)
                    schema.Required = new List<string>();
                foreach (var property in nonNullableProperties)
                {
                    if (!schema.Required.Contains(property))
                        schema.Required.Add(property);
                }
            }
        }
    }

**Comment by ntregillus** [comment]
Interesting... are you not using the DataAnnotations for validation? How do you know if the client calling your service passed zero, or if the request deserialization defaulted to zero? while your Swashbuckle filter will mark the variable as required in the swagger, there is nothing stopping an integrator from passing an invalid request that is missing the IntRequired value. Passing a request to your api, the model would deserialize to `IntRequired` = 0.

I would recommend your models your api is accepting always has nullable fields, but the required fields are denoted with the Required attribute. that way you can be sure Zero for you int variable always was what the client provided, not what the serializer defaulted to setting to your non null field.

**Comment by ThisNoName** [comment]
Thanks for the follow up.  I was referring to return model, not request.  By default, all returned DateTime, int, boolean become optional in swagger spec, unless you marked them [Required].  Which is quite a chore, especially the underlying code already differentiate between int and int?

**Comment by domaindrivendev** [comment]
Please note that in JSON Schema, “nullable” does not  imply “not required”. Those two attributes of a JSON primitive are unrelated.

For example, a field can be required and allow null as a value:

{ “foobar”: null }

By design, SB does not make assumptions and therefore does not mark use int vs int? do determine if a field should be marked required or not

**Comment by ThisNoName** [comment]
The problem arise when you try to generate C# client based on the spec.  It goes like this:

    c# =>           public Datetime foobar {get;set;}    
    swashbuckle =>  {"foobar" : null }
    autorest =>     public DateTime? foobar {get;set;}

Now when trying to consume the generated client, I can't do simple things like

    foobar.ToString("MM/dd/yyyy"}

Instead,

    foobar.HasValue ? foobar.value.ToString("MM/dd/yyyy") : ""

Even though I know foobar always return a valid date.  Whether we leave the hassle to the client, or add [Required] on all properties, it becomes quite cumbersome when there're a bunch of date, int and boolean in the mix.   

I read Swashbuckle was following some standard that lead to this.  But it does appear to be a legitimate hassle for lots of people.  Wondering if someone already ported the original [AutoRestExtension](https://github.com/Alegrowin/Swashbuckle.AutoRestExtensions), or maybe this can be built into Swashbuckle as an option?  

I know NSwagStudio has this built it, the specs it generates always follow the original C# declared type, like Datetime, Datetime?, int, int?, etc.

Thanks.

**Comment by zhangb99** [comment]
Eureka! 

This

    public class NullableTypeSchemaFilter : ISchemaFilter  {
        public void Apply(Schema model, SchemaFilterContext context)   {
            if (context.SystemType.IsValueType 
                    && Nullable.GetUnderlyingType(context.SystemType) == null)
                model.Extensions.Add("x-nullable", false);
      } }

and that

    services.AddSwaggerGen(c =>  {
              c.SchemaFilter<NullableTypeSchemaFilter>();
    });

And wa-la!  x-nullable:false for non-nullable int, datetime, bool, etc., without cluttering up rest of the specs.  

Maybe this can be one of the standard filters, or an option in SweggerGen?  Thanks for the help.

**Comment by ThisNoName** [comment]
Eureka!

This
```
public class NullableTypeSchemaFilter : ISchemaFilter  {
    public void Apply(Schema model, SchemaFilterContext context)   {
        if (context.SystemType.IsValueType 
                && Nullable.GetUnderlyingType(context.SystemType) == null)
            model.Extensions.Add("x-nullable", false);
  } }
```
and that
```
services.AddSwaggerGen(c =>  {
          c.SchemaFilter<NullableTypeSchemaFilter>();
});
```
And wa-la! x-nullable:false for non-nullable int, datetime, bool, etc., without cluttering up rest of the specs.

Maybe this can be one of the standard filters, or an option in SweggerGen? Thanks for the help.

**Comment by coderob** [comment]
The Swagger doc generation seems to only respects the "required" attribute for nullable types. Is this a known bug, or am I missing something?

> .NET Core 1.1.1
> "Swashbuckle.AspNetCore" Version="2.5.0"
> "Swashbuckle.AspNetCore.SwaggerGen" Version="2.5.0" 
> "Swashbuckle.AspNetCore.SwaggerUI" Version="2.5.0"

e.g.:
```
public class ProfileUpdateInbound
{
    [Required]
    public long? ProfileId { get; set; }

    [Required]
    public int UnicornId { get; set; }
    ...
}

[HttpPut]
public void Update (ProfileUpdateInbound profile) 
{
    ... 
}
```

![image](https://user-images.githubusercontent.com/1413487/46711892-5c085280-cc1c-11e8-9333-5d90032c3ebc.png)

**Comment by coderob** [comment]
> The Swagger doc generation seems to only respects the "required" attribute for nullable types. Is this a known bug, or am I missing something?

I believe I just found the answer that it was a bug and that it was [fixed in v3.0.0](https://github.com/domaindrivendev/Swashbuckle.AspNetCore/releases/tag/v3.0.0) 💯 Nice docs!
